Isofibrations

nimarasekh opened the pull request:
Definition of Isofibrations modeled based on trivial fibrations along with @mpenciak
Does one need to also adjust other files (like stuff related to blueprint, ... )?

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/SimplicialSet/MorphismProperty.lean:
```diff
@@ -26,4 +27,35 @@ def trivialFibration : MorphismProperty SSet := fun _ _ p ↦ BoundaryInclusions
+  | mk (n i : ℕ) (low : 0 < i) (high : i < n) : InnerInclusion (hornInclusion n i)
+
+/-- The class of inner horn inclusions as a MorphismProperty -/
+def InnerInclusions : MorphismProperty SSet := fun _ _ p ↦ InnerInclusion p
```
For readability later, perhaps these should be called "InnerHornInclusions"? (I'm unclear about capitalization; see discussion on zulip.)

nimarasekh replied:
Okay, I followed the discussion on Zulip, and I am still unsure what the capitalization should be, so I'll keep that as is for now and just adjust the name as suggested.

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/SimplicialSet/MorphismProperty.lean:
```diff
@@ -26,4 +27,35 @@ def trivialFibration : MorphismProperty SSet := fun _ _ p ↦ BoundaryInclusions
+def InnerInclusions : MorphismProperty SSet := fun _ _ p ↦ InnerInclusion p
+
+/-- Inductive definition of being equal to the coherent iso -/
+inductive CoherentIsoInclusion : {X Y : SSet} → (X ⟶ Y) → Prop
```
We'll want to be able to lift against the inclusion of either 0 or 1. Perhaps it's best to have both as elements of this class? 

Then (later) you could prove that having the rlp against `coherentIso.pt .zero` implies the same for `coherentIso.pt .one` by a retract argument. Then you could have a lemma that helps you define an isofibration from either lifting property.

nimarasekh replied:
> We'll want to be able to lift against the inclusion of either 0 or 1. Perhaps it's best to have both as elements of this class?
> 
> Then (later) you could prove that having the rlp against `coherentIso.pt .zero` implies the same for `coherentIso.pt .one` by a retract argument. Then you could have a lemma that helps you define an isofibration from either lifting property.

Sure, I guess my thinking was to minimize the requirements to define it to make constructing it easier, but then of course using it is more difficult.

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/SimplicialSet/MorphismProperty.lean:
```diff
@@ -26,4 +27,35 @@ def trivialFibration : MorphismProperty SSet := fun _ _ p ↦ BoundaryInclusions
+
+/-- Definition of isofibration: An isofibration of quasi-categories is a morphism that
+has the rlp wrt inner horn inclusions and the inclusion into the walking coherent iso. -/
+def isoFibration : MorphismProperty SSet := fun _ _ p ↦ InnerIsoInclusions.rlp p
```
I only like this definition when the ambient simplicial sets are quasi-categories (suffices, and I think is better, to assume this just for the codomain). Can that be added to the morphism property api somehow? (If not, ask @mckoen and others about this on zulip).

nimarasekh replied:
I think if there is a category of quasi-categories, then it should be possible to use MorphismProperty directly, however, if you want just one of them, then yeah I probably would have to ask.

nimarasekh replied:
FYI @emilyriehl the definition you mentioned here, does not match with the latex file, where it says "between quasi-categories". Should that be adjusted accordingly?

emilyriehl replied:
I like the solution of stating this definition for the category of quasi-categories only. Then we can apply the morphism property api directly without modification. Would you mind adding that subcategory for us, perhaps called QCat? It will be useful anyway. (NB: I'm pretty sure Lean knows about full subcategories.)

emilyriehl commented:
@pitmonticone can you tell if the build_project check is functioning as intended here?

pitmonticone commented:
@emilyriehl Yes, it’s working precisely as expected. As you can see merging isn’t possible without status check!

emilyriehl commented:
So if I (as reviewer) want to perform this status check before approving a merge, what exactly do I need to do? Or is this something done by the PR authors? Or done automatically?

emilyriehl commented:
> Definition of Isofibrations modeled based on trivial fibrations along with @mpenciak Does one need to also adjust other files (like stuff related to blueprint, ... )?

@nimarasekh it would be great actually if you would update this for me, so I don't have to do it myself. If you've defined a term that is related to one of the bubbles in to the dependency graph that means it's mentioned explicitly in the blueprint. What you need to add to the blueprint is two bits of code 
```
\leanok
\lean{full-name-of-term-including-namespaces-here}
```
See the blueprint for lots of examples and let me know if you need help. Speaking of which, perhaps we put Isofibration in a namespace like `SSet/Quasicategory`?

pitmonticone commented:
Oh, now I see what you mean @emilyriehl (sorry, I’m from mobile). 

The additional “build_project” isn’t necessary. Could you remove it for testing purposes?

pitmonticone commented:
Once you have done that, can you tell me the list of available checks to select?

nimarasekh commented:
> > Definition of Isofibrations modeled based on trivial fibrations along with @mpenciak Does one need to also adjust other files (like stuff related to blueprint, ... )?
> 
> @nimarasekh it would be great actually if you would update this for me, so I don't have to do it myself. If you've defined a term that is related to one of the bubbles in to the dependency graph that means it's mentioned explicitly in the blueprint. What you need to add to the blueprint is two bits of code
> 
> ```
> \leanok
> \lean{full-name-of-term-including-namespaces-here}
> ```
> 
> See the blueprint for lots of examples and let me know if you need help. Speaking of which, perhaps we put Isofibration in a namespace like `SSet/Quasicategory`?

Okay I think I made the content changes you suggested and it's now in the PR.

nimarasekh commented:
Okay, @emilyriehl I corrected the definition of isofibrations in the latex file (removing the QCat condition on the domain) The Lean definition is also adjusted. However, I think the solution is not very elegant cause I added a new type of morphismproperty for that.

emilyriehl commented:
Sorry to be slow @nimarasekh. I think your initial suggestion (just to implement this definition in the subcategory of quasi-categories) is the more elegant. Would you mind doing that?

emilyriehl commented on blueprint/src/content.tex:
```diff
@@ -505,7 +505,9 @@ \section{Simplicial sets}\label{sec:simplicial-sets}
 
 \begin{definition}[isofibration]\label{defn:qcat-isofibration}
   \uses{defn:coherent-isomorphism}
-  A simplicial map $f \colon A \to B$ between quasi-categories is an \textbf{isofibration} if it lifts against the inner horn inclusions, as displayed below-left, and also against the inclusion of either vertex into the free-living isomorphism $\iso$.
```
Let's either revert this change or add a footnote to clarify. You and I know it is more efficient but I worry it won't be clear to everyone that $A$ is necessarily also a quasi-category. You can pick which.

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/SimplicialSet/MorphismProperty.lean:
```diff
@@ -4,14 +4,24 @@ Released under Apache 2.0 license as described in the file LICENSE.
+section morphismPropertyTargetQCat
+
+/-- A variation of `MorphismProperty` which assumes the target is a quasicategory. -/
+def QCatTargetMorphismProperty :=
```
Let's cut this for now (unless you disagree).

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/SimplicialSet/MorphismProperty.lean:
```diff
@@ -26,4 +36,47 @@ def trivialFibration : MorphismProperty SSet := fun _ _ p ↦ BoundaryInclusions
+/-- Definition of isofibration: A morphism of simplicial sets with target a quasi-category 
+  is an isofibration if it satisfies the rlp with wrt the inner horn inclusions and the two 
+  endpoint inclusion into the walking coherent iso. -/
+def isoFibration : QCatTargetMorphismProperty := fun _ _ p ↦ InnerIsoInclusions.rlp p
```
I think this should be capitalized (and changed to be a term of type `MorphismProperty QCat`.

nimarasekh replied:
Does that mean the term "trivialFibration" above should also be capitalized?

emilyriehl replied:
I guess?

emilyriehl commented:
Thanks @nimarasekh and @mpenciak for all the edits. Do you want to define the full subcategory of quasi-categories so that we can type `Isofibration` as `Isofibration : MorphismProperty QCat`? Or would you rather we merge this first and then have someone else do that later? (Eg, my seminar could make this change this afternoon.)

nimarasekh commented:
> Thanks @nimarasekh and @mpenciak for all the edits. Do you want to define the full subcategory of quasi-categories so that we can type `Isofibration` as `Isofibration : MorphismProperty QCat`? Or would you rather we merge this first and then have someone else do that later? (Eg, my seminar could make this change this afternoon.)

I think we are not done yet and currently working on this. I guess one question would be where should we define the category QCat? I think it should be in the quasicategory file, but there is none in formathlib, right? So you have a suggestion how to proceed?

emilyriehl commented:
Yes: in that circumstance you should create a new "Quasicategory" file in our "ForMathlib" in the same location as the parallel file in Mathlib (eg under "AlgebraicTopology/SimplicialSets"). Start that file by importing the mathlib quasi-categories file. Then add your addition.

I'm hoping this structure will 
(i) make it easy to later PR whatever you have added into mathlib. 
(ii) make it easy for us to understand what is and is not in mathlib.

nimarasekh commented:
> Yes: in that circumstance you should create a new "Quasicategory" file in our "ForMathlib" in the same location as the parallel file in Mathlib (eg under "AlgebraicTopology/SimplicialSets"). Start that file by importing the mathlib quasi-categories file. Then add your addition.
> 
> I'm hoping this structure will (i) make it easy to later PR whatever you have added into mathlib. (ii) make it easy for us to understand what is and is not in mathlib.

Okay, then we'll proceed as suggested and add it to the PR. If we encounter any issues, then we'll reach out here or via Zulip.

nimarasekh commented:
Okay @emilyriehl you can check it now, we should be done.

emilyriehl commented:
Bon merge!
